File: chapter2/f_anova.py
# Import libraries
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import f_oneway, f

# Random seed
np.random.seed(1234)

# Formulate hypothesis
# H0: All data comes from same distribution
# H1: At least one sample comes from different distribution

# Generate gaussian observations
group1 = np.random.normal(loc = 5, scale = 1, size = 10)
group2 = np.random.normal(loc = 8, scale = 1, size = 10)
group3 = np.random.normal(loc = 5, scale = 1, size = 10)
data = [group1, group2, group3]
print("Observations:\ngroup1: ", data[0], "\ngroup2: ", data[1], "\ngroup3: ", data[2])

# Plot observations
plt.figure(figsize = (8, 6))
plt.violinplot(data, showmeans = True, showextrema = True)
plt.title("Plot of observations")
plt.ylabel("Sample mean")
plt.show()

# Compute group means
mean1, mean2, mean3 = np.mean(data[0]), np.mean(data[1]), np.mean(data[2])
overall_mean = np.mean(np.concatenate(data))
print("\nmean1: ", mean1, "\nmean2: ", mean2, "\nmean3: ", mean3)

# Degrees of freedom
df_between = len(data) - 1
df_within = len(group1) + len(group2) + len(group3) - len(data)

# Variation between groups (SSB)
ssb = (len(group1) * ((mean1 - overall_mean))**2 + 
    len(group2) * ((mean2 - overall_mean))**2 +
    len(group3) * ((mean3 - overall_mean))**2)

# Variation within groups (SSW)
ssw = np.sum((group1 - mean1)**2) + np.sum((group2 - mean2)**2) + np.sum((group3 - mean3)**2)

# Normalize with degrees of freedom
ssb = ssb / df_between
ssw = ssw / df_within

# F-statistic manual calculation
F_stat_manual = ssb / ssw
p_value_manual = 1 - f.cdf(F_stat_manual, df_between, df_within)
print("\nF ANOVA (manual): ", F_stat_manual)
print("p-value (manual): ", p_value_manual)

# F-statistic with scipy library
F_stat_scipy, p_value_scipy = f_oneway(group1, group2, group3)
print("\nF ANOVA (scipy): ", F_stat_scipy)
print("p-value (scipy): ", p_value_scipy)

File: chapter2/exercises/f_anova.R
# Import libraries
library(ggplot2)
library(gridExtra)

# Set random seed
set.seed(42)

# Generate gaussian observations
group1 <- rnorm(10, mean = 5, sd = 1)
group2 <- rnorm(10, mean = 5, sd = 1)
group3 <- rnorm(10, mean = 5, sd = 1)
data <- list(group1, group2, group3)

# Plot observations
df <- data.frame(
  value = c(group1, group2, group3),
  group = factor(rep(1:3, each = 10))
)

p <- ggplot(df, aes(x = group, y = value)) +
  geom_violin(trim = FALSE, fill = "lightblue") +
  stat_summary(fun = mean, geom = "point", shape = 4, size = 3, color = "red") +
  labs(title = "Plot of observations", y = "Sample mean") +
  theme_minimal()

print(p)

# Compute group means
mean1 <- mean(group1)
mean2 <- mean(group2)
mean3 <- mean(group3)
overall_mean <- mean(c(group1, group2, group3))

# Compute sum of squares between (SSB) and within (SSW)
ssb <- length(data) * ((mean1 - overall_mean)^2 + (mean2 - overall_mean)^2 + (mean3 - overall_mean)^2)
ssw <- sum((group1 - mean1)^2) + sum((group2 - mean2)^2) + sum((group3 - mean3)^2)
df_between <- length(data) - 1  # n - 1
df_within <- length(group1) + length(group2) + length(group3) - length(data)  # N - n

# Mean Squares
msb <- ssb / df_between
msw <- ssw / df_within
df1 <- length(data)
df2 <- length(group1) + length(group2) + length(group3) - length(data)

# F-statistic
F_stat_manual <- msb / msw
p_value_manual <- 1 - pf(F_stat_manual, df1, df2)

cat("\nANOVA (manual):\n")
cat(sprintf("F statistic (manual): %.2f\n", F_stat_manual))
cat(sprintf("p value (manual): %.4f\n", p_value_manual))

# Compute ANOVA (built-in function)
anova_res <- aov(value ~ group, data = df)
summary_res <- summary(anova_res)
F_stat_scipy <- summary_res[[1]]$`F value`[1]
p_value_scipy <- summary_res[[1]]$`Pr(>F)`[1]

cat("\nANOVA (R built-in):\n")
cat(sprintf("F statistic (R built-in): %.2f\n", F_stat_scipy))
cat(sprintf("p value (R built-in): %.4f\n", p_value_scipy))

# Conclusion based on the p-value
alpha <- 0.05
if (p_value_manual < alpha) {
  cat("\nReject the null hypothesis: There are significant differences between the groups.\n")
} else {
  cat("\nFail to reject the null hypothesis: No significant differences between the groups.\n")
}
